### PR TITLE
Fix for Bisa/factorio-init#194: Use factorio api for latest-releases

### PR DESCRIPTION
**Pull request by Legrems**

### CODE REVIEW

File: factorio [hunk 1]
```diff
@@ -465,24 +465,25 @@ function install(){
...
+      LATEST_RELEASES_JSON_KEY=".stable.headless"
+    fi
+
+    # We fetch the version in the json based on the LATEST_RELEASES_JSON_KEY
+    if ! version="$(echo $httpresponse | jq -r $LATEST_RELEASES_JSON_KEY)"; then
+        info "${version}"
+        error "Aborting install, unable to find version with key ${LATEST_RELEASES_JSON_KEY}"
+    fi
+
+    filename="factorio_headless_x64_${version}.tar.xz"
```
Comment (Legrems) [inline factorio:486]: I'm using this to keep the same filename as the script uses currently.

**Comment by tmzasz** [comment]
theres 2 methods that do not require the user to install new software nor do they require a large change to the check core.

Method 1: grep + awk
curl -LIs https://www.factorio.com/get-download/stable/headless/linux64 | grep Location | awk -F "?" '{print $1}' | awk -F "/" '{print $NF}' | awk -F "_" '{print $4}' | awk -F ".tar" '{print $1}'

Returns something like this as of today : 1.1.57

Method 2: grep + awk on new API site above

to get experimental
curl -s https://factorio.com/api/latest-releases | awk -F ":" '{print $5 $9}' | awk -F "\"" '{print $2}'
outputs 1.1.57
stable
curl -s https://factorio.com/api/latest-releases | awk -F ":" '{print $5 $9}' | awk -F "\"" '{print $6}'
outputs 1.1.57

both methods are simpler and do not force the user to install/rely on more external software

**Review by tmzasz — CHANGES_REQUESTED** [review]
use grep + awk or just awk even and do not require the user to install more software just to handle json

there's 2 methods that do not require the user to install new software nor do they require a large change to the check core.

### Method 1: grep + awk
curl -LIs https://www.factorio.com/get-download/stable/headless/linux64 | grep Location | awk -F "?" '{print $1}' | awk -F "/" '{print $NF}' | awk -F "_" '{print $4}' | awk -F ".tar" '{print $1}'

Returns something like this as of today : 1.1.57

**to get the file name you can do this**
curl -LIs https://www.factorio.com/get-download/stable/headless/linux64 | grep Location | awk -F "?" '{print $1}' | awk -F "/" '{print $NF}'

which returns this
factorio_headless_x64_1.1.57.tar.xz

### Method 2: awk on new API site

**to get experimental**
_curl -s https://factorio.com/api/latest-releases | awk -F ":" '{print $5 $9}' | awk -F """ '{print $2}'_

outputs 1.1.57

**to get stable**
_curl -s https://factorio.com/api/latest-releases | awk -F ":" '{print $5 $9}' | awk -F """ '{print $6}'_

outputs 1.1.57

both methods are simpler and do not force the user to install/rely on more external software

for Method 2 this is the sample raw output that awk works on
{"experimental":{"alpha":"1.1.57","demo":"1.1.57","headless":"1.1.57"},"stable":{"alpha":"1.1.57","demo":"1.1.57","headless":"1.1.57"}}
what its doing is breaking down the response first by : parts and then by a " and then returning the numbered column

the benefit is that any system capable of running a factorio server should also have awk/grep installed by default as they are core POSIX functions

**Comment by Legrems** [comment]
> use grep + awk or just awk even and do not require the user to install more software just to handle json
> 
> there's 2 methods that do not require the user to install new software nor do they require a large change to the check core.
> ### Method 1: grep + awk
> 
> _curl -LIs https://www.factorio.com/get-download/stable/headless/linux64 | grep Location | awk -F "?" '{print $1}' | awk -F "/" '{print $NF}' | awk -F "_" '{print $4}' | awk -F ".tar" '{print $1}'_
> 
> Returns something like this as of today : 1.1.57
> 
> **to get the file name you can do this** _curl -LIs https://www.factorio.com/get-download/stable/headless/linux64 | grep Location | awk -F "?" '{print $1}' | awk -F "/" '{print $NF}'_
> 
> which returns this factorio_headless_x64_1.1.57.tar.xz
> ### Method 2: awk on new API site
> 
> **to get experimental** _curl -s https://factorio.com/api/latest-releases | awk -F ":" '{print $5 $9}' | awk -F """ '{print $2}'_
> 
> outputs 1.1.57
> 
> **to get stable** _curl -s https://factorio.com/api/latest-releases | awk -F ":" '{print $5 $9}' | awk -F """ '{print $6}'_
> 
> outputs 1.1.57
> 
> both methods are simpler and do not force the user to install/rely on more external software
> 
> for Method 2 this is the sample raw output that awk works on {"experimental":{"alpha":"1.1.57","demo":"1.1.57","headless":"1.1.57"},"stable":{"alpha":"1.1.57","demo":"1.1.57","headless":"1.1.57"}} what its doing is breaking down the response first by : parts and then by a " and then returning the numbered column
> 
> the benefit is that any system capable of running a factorio server should also have awk/grep installed by default as they are core POSIX functions

The awk method is not robust to any changes made to the latest releases API.

The advantage of jq (which is by the way super light and you should consider using this for parsing json) is that you rely on key from the json instead of arbitrary number of ":".

Any server capable of running factorio is also capable of having jq installed, since it has no runtime dependencies.

**Comment by tmzasz** [comment]
the advantage of awk is still 1 less software to install. There is no need to keep installing more and more packages to do something the os is already capable of doing by default. Less complexity is always better than more.

**Comment by llann** [comment]
I don't understand that "fear" to install such a small lib on a system knowing that we install Factorio and everything that goes with it. Using this makes the code more readable and more resistant to changes that the dev. may do in the API.

**Comment by tmzasz** [comment]
its not a "fear" its a principle, why install more and more software to do tasks that the currently installed software can already handle just fine.

It would be like installing a distributed calculation program to handle 1+1 which the os's calculator can already do ( extreme example but valid example )

**Comment by tmzasz** [comment]
theres also the issue of slowly creeping into Dependency hell for nothing

**Comment by Legrems** [comment]
"Creeing into dependency hell for nothing"
Ok, then don't install factorio-init (which by the way don't work atm, so nice dependancy) to "just" install a server because it's still one more dependances than just downloading and launching the game.

**Comment by Legrems** [comment]
Just before definetly closing this, i'm curious to know your opinion on this @Bisa.

Thank's in advance !

**Comment by Bisa** [comment]
Thnx for the suggested fix @Legrems - I did however opt to use already existing tools, at some point in the future I may look at using the api, but much like @tmzasz I try to stick to as few "extras" as I can (when I remember to and when I do not whirr off into some random "this is so cool" spree ;) )

**Pull request closed without merging**